File: pipeline/sun.py
#!/usr/bin/env python3
"""Solar position (NOAA algorithm), dependency-free.

solar_position(dt_utc, lat, lon) -> (azimuth_deg_from_N_clockwise, elevation_deg, declination_deg).
geneva_utc(...) converts a Geneva wall-clock time to UTC (DST-aware via zoneinfo).
The TypeScript twin in src/routing/sun.ts MUST match this within ~0.5 deg.
"""
import math
from datetime import datetime
from zoneinfo import ZoneInfo

UTC = ZoneInfo("UTC")


def geneva_utc(year, month, day, hour, minute=0, tz="Europe/Zurich"):
    return datetime(year, month, day, hour, minute, tzinfo=ZoneInfo(tz)).astimezone(UTC)


def _julian_day(dt):
    y, m = dt.year, dt.month
    d = dt.day + (dt.hour + dt.minute / 60 + dt.second / 3600) / 24.0
    if m <= 2:
        y -= 1
        m += 12
    a = y // 100
    b = 2 - a + a // 4
    return int(365.25 * (y + 4716)) + int(30.6001 * (m + 1)) + d + b - 1524.5


def _clamp(x, lo=-1.0, hi=1.0):
    return max(lo, min(hi, x))


def solar_position(dt_utc, lat, lon):
    """lat/lon in degrees (east positive). dt_utc tz-aware (UTC)."""
    jd = _julian_day(dt_utc)
    jc = (jd - 2451545.0) / 36525.0
    l0 = (280.46646 + jc * (36000.76983 + jc * 0.0003032)) % 360.0
    m = 357.52911 + jc * (35999.05029 - 0.0001537 * jc)
    e = 0.016708634 - jc * (0.000042037 + 0.0000001267 * jc)
    mr = math.radians(m)
    c = (math.sin(mr) * (1.914602 - jc * (0.004817 + 0.000014 * jc))
         + math.sin(2 * mr) * (0.019993 - 0.000101 * jc)
         + math.sin(3 * mr) * 0.000289)
    true_long = l0 + c
    omega = 125.04 - 1934.136 * jc
    app_long = true_long - 0.00569 - 0.00478 * math.sin(math.radians(omega))
    obliq0 = 23 + (26 + (21.448 - jc * (46.815 + jc * (0.00059 - jc * 0.001813))) / 60) / 60
    obliq = obliq0 + 0.00256 * math.cos(math.radians(omega))
    decl = math.degrees(math.asin(_clamp(math.sin(math.radians(obliq)) * math.sin(math.radians(app_long)))))
    y = math.tan(math.radians(obliq / 2)) ** 2
    eot = 4 * math.degrees(
        y * math.sin(2 * math.radians(l0)) - 2 * e * math.sin(mr)
        + 4 * e * y * math.sin(mr) * math.cos(2 * math.radians(l0))
        - 0.5 * y * y * math.sin(4 * math.radians(l0))
        - 1.25 * e * e * math.sin(2 * mr))
    mins = dt_utc.hour * 60 + dt_utc.minute + dt_utc.second / 60.0
    tst = (mins + eot + 4 * lon) % 1440.0
    ha = tst / 4.0 - 180.0
    if ha < -180:
        ha += 360.0
    lat_r, decl_r, ha_r = math.radians(lat), math.radians(decl), math.radians(ha)
    cos_zen = _clamp(math.sin(lat_r) * math.sin(decl_r) + math.cos(lat_r) * math.cos(decl_r) * math.cos(ha_r))
    zen = math.degrees(math.acos(cos_zen))
    el = 90.0 - zen
    denom = math.cos(lat_r) * math.sin(math.radians(zen))
    if abs(denom) < 1e-9:
        az = 0.0 if lat > decl else 180.0
    else:
        cos_az = _clamp((math.sin(lat_r) * math.cos(math.radians(zen)) - math.sin(decl_r)) / denom)
        a = math.degrees(math.acos(cos_az))
        az = (a + 180.0) % 360.0 if ha > 0 else (540.0 - a) % 360.0
    return az, el, decl


if __name__ == "__main__":
    # sanity: Geneva summer, morning sun in the east, noon near south
    lat, lon = 46.20, 6.14
    for h in (8, 12, 16):
        az, el, decl = solar_position(geneva_utc(2026, 6, 21, h), lat, lon)
        print(f"Geneva 2026-06-21 {h:02d}:00 local -> az {az:6.1f}  el {el:5.1f}  decl {decl:5.2f}")

File: src/routing/sun.ts
// Solar position (NOAA algorithm) — TypeScript twin of pipeline/sun.py.
// Drives the sun indicator and time-bin selection. Must match sun.py within ~0.5 deg.
// Geneva reference; Europe/Zurich DST handled explicitly (EU rule).

const LAT = 46.20, LON = 6.14;
const D2R = Math.PI / 180, R2D = 180 / Math.PI;
const clamp = (x: number, lo = -1, hi = 1) => (x < lo ? lo : x > hi ? hi : x);

// last Sunday (UTC day-of-month) of a 0-indexed month
function lastSunday(year: number, monthIdx0: number): number {
  const d = new Date(Date.UTC(year, monthIdx0 + 1, 0)); // last day of the month
  return d.getUTCDate() - d.getUTCDay();
}
// Europe/Zurich offset in hours for a Geneva wall-clock date: CET=+1, CEST=+2
function zurichOffset(y: number, m: number, d: number): number {
  const marL = lastSunday(y, 2), octL = lastSunday(y, 9);   // last Sun Mar / Oct
  const onOrAfter = (mm: number, dd: number, bm: number, bd: number) => mm > bm || (mm === bm && dd >= bd);
  const dst = onOrAfter(m, d, 3, marL) && !onOrAfter(m, d, 10, octL);
  return dst ? 2 : 1;
}

/** A Geneva wall-clock (y, m=1..12, d, hour) -> the UTC instant. */
export function genevaToUTC(y: number, m: number, d: number, hour: number): Date {
  return new Date(Date.UTC(y, m - 1, d, hour - zurichOffset(y, m, d), 0, 0));
}

function julianDay(utc: Date): number {
  let y = utc.getUTCFullYear(), mo = utc.getUTCMonth() + 1;
  const day = utc.getUTCDate() + (utc.getUTCHours() + utc.getUTCMinutes() / 60 + utc.getUTCSeconds() / 3600) / 24;
  if (mo <= 2) { y -= 1; mo += 12; }
  const A = Math.floor(y / 100), B = 2 - A + Math.floor(A / 4);
  return Math.floor(365.25 * (y + 4716)) + Math.floor(30.6001 * (mo + 1)) + day + B - 1524.5;
}

export interface Sun { az: number; el: number; decl: number; }

export function solarPosition(utc: Date, lat = LAT, lon = LON): Sun {
  const jc = (julianDay(utc) - 2451545.0) / 36525.0;
  const l0 = ((280.46646 + jc * (36000.76983 + jc * 0.0003032)) % 360 + 360) % 360;
  const m = 357.52911 + jc * (35999.05029 - 0.0001537 * jc);
  const e = 0.016708634 - jc * (0.000042037 + 0.0000001267 * jc);
  const mr = m * D2R;
  const c = Math.sin(mr) * (1.914602 - jc * (0.004817 + 0.000014 * jc))
    + Math.sin(2 * mr) * (0.019993 - 0.000101 * jc) + Math.sin(3 * mr) * 0.000289;
  const trueLong = l0 + c;
  const omega = 125.04 - 1934.136 * jc;
  const appLong = trueLong - 0.00569 - 0.00478 * Math.sin(omega * D2R);
  const obliq0 = 23 + (26 + (21.448 - jc * (46.815 + jc * (0.00059 - jc * 0.001813))) / 60) / 60;
  const obliq = obliq0 + 0.00256 * Math.cos(omega * D2R);
  const decl = Math.asin(clamp(Math.sin(obliq * D2R) * Math.sin(appLong * D2R))) * R2D;
  const y = Math.tan(obliq / 2 * D2R) ** 2;
  const eot = 4 * R2D * (y * Math.sin(2 * l0 * D2R) - 2 * e * Math.sin(mr)
    + 4 * e * y * Math.sin(mr) * Math.cos(2 * l0 * D2R)
    - 0.5 * y * y * Math.sin(4 * l0 * D2R) - 1.25 * e * e * Math.sin(2 * mr));
  const mins = utc.getUTCHours() * 60 + utc.getUTCMinutes() + utc.getUTCSeconds() / 60;
  const tst = (mins + eot + 4 * lon) % 1440;
  let ha = tst / 4 - 180; if (ha < -180) ha += 360;
  const latR = lat * D2R, declR = decl * D2R;
  const cosZen = clamp(Math.sin(latR) * Math.sin(declR) + Math.cos(latR) * Math.cos(declR) * Math.cos(ha * D2R));
  const zen = Math.acos(cosZen) * R2D;
  const el = 90 - zen;
  const denom = Math.cos(latR) * Math.sin(zen * D2R);
  let az: number;
  if (Math.abs(denom) < 1e-9) az = lat > decl ? 0 : 180;
  else {
    const a = Math.acos(clamp((Math.sin(latR) * Math.cos(zen * D2R) - Math.sin(declR)) / denom)) * R2D;
    az = ha > 0 ? (a + 180) % 360 : (540 - a) % 360;
  }
  return { az, el, decl };
}

/** Sun for a Geneva wall-clock time. */
export function sunForGeneva(y: number, m: number, d: number, hour: number): Sun {
  return solarPosition(genevaToUTC(y, m, d, hour));
}

/** Current Geneva wall-clock parts (DST-correct via Intl). */
export function genevaNow(): { y: number; m: number; d: number; h: number } {
  const f = new Intl.DateTimeFormat("en-CA", {
    timeZone: "Europe/Zurich", year: "numeric", month: "2-digit", day: "2-digit",
    hour: "2-digit", hourCycle: "h23",
  });
  const o: Record<string, string> = {};
  for (const p of f.formatToParts(new Date())) o[p.type] = p.value;
  return { y: +o.year, m: +o.month, d: +o.day, h: +o.hour };
}

const CARD = ["N", "NE", "E", "SE", "S", "SO", "O", "NO"];
export function cardinal(az: number): string {
  return CARD[Math.round(az / 45) % 8];
}
